File: src/java/org/codehaus/testdox/intellij/config/ConfigurationController.java
package org.codehaus.testdox.intellij.config;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.ProjectComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.options.ConfigurableProvider;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.project.Project;
import com.intellij.util.xmlb.XmlSerializerUtil;
import org.codehaus.testdox.intellij.IconHelper;
import org.codehaus.testdox.intellij.TemplateNameResolver;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Collections;
import java.util.List;

@State(
        name = ConfigurationController.COMPONENT_NAME,
        storages = {
                @Storage(
                        id = ConfigurationController.COMPONENT_NAME,
                        file = "$PROJECT_FILE$"
                )}
)

public class ConfigurationController implements ProjectComponent, Configurable, PersistentStateComponent<ConfigurationBean> {

    static final String COMPONENT_NAME = "TestDoxConfiguration";

    protected ConfigurationUI panel;

    private final ConfigurationBean bean = new ConfigurationBean();

    public ConfigurationController() {
        setDefaults();
    }

    // BaseComponent ---------------------------------------------------------------------------------------------------

    public void initComponent() {
    }

    public void projectOpened() {
    }

    public void projectClosed() {
    }

    public void disposeComponent() {
    }

    @NotNull
    public String getComponentName() {
        return COMPONENT_NAME;
    }

    private void setDefaults() {
        if (bean.getCustomPackages() == null) {
            bean.setCustomPackages(Collections.<String>emptyList());
        }
        if (bean.getTestNameTemplate() == null) {
            bean.setTestNameTemplate(TemplateNameResolver.DEFAULT_TEMPLATE);
        }
        if (bean.getTestMethodPrefix() == null && bean.getTestMethodAnnotation() == null) {
            bean.setTestMethodPrefix(TemplateNameResolver.DEFAULT_PREFIX);
        }
        bean.setCreateTestIfMissing(true);
        bean.setAutoApplyChangesToTests(true);
        bean.setDeletePackageOccurrences(true);
    }

    // UnnamedConfigurable ---------------------------------------------------------------------------------------------

    public void apply() throws ConfigurationException {
        bean.setAllowCustomPackages(panel.getCustomMappingStatus());
        bean.setCustomPackages(panel.getCustomPackageMappings());
        bean.setTestNameTemplate(panel.getTestNameTemplate());
        bean.setCreateTestIfMissing(panel.getCreateTestIfMissing());
        bean.setUnderscoreMode(panel.getUseUnderscore());
        bean.setShowFullyQualifiedClassName(panel.getShowFullyQualifiedClassName());
        bean.setAutoApplyChangesToTests(panel.getAutoApplyChangesToTest());
        bean.setDeletePackageOccurrences(panel.getDeletePackageOccurrences());
        String prefix = panel.getTestMethodPrefix();
        if (prefix != null && prefix.startsWith("@")) {
            bean.setTestMethodPrefix(null);
            bean.setTestMethodAnnotation(prefix);
            bean.setUsingAnnotations(true);
        } else {
            bean.setTestMethodPrefix(prefix);
            bean.setTestMethodAnnotation(null);
            bean.setUsingAnnotations(false);
        }
    }

    public JComponent createComponent() {
        panel = new ConfigurationPanel();
        reset();
        return (ConfigurationPanel) panel;
    }

    public void disposeUIResources() {
        panel = null;
    }

    public String getDisplayName() {
        return "TestDox";
    }

    public String getHelpTopic() {
        return "";
    }

    public Icon getIcon() {
        return IconHelper.getIcon(IconHelper.TESTDOX_ICON);
    }

    public boolean isModified() {
        return bean.isAllowCustomPackages() != panel.getCustomMappingStatus()
                || !equals(bean.getCustomPackages(), panel.getCustomPackageMappings())
                || !bean.getTestNameTemplate().equals(panel.getTestNameTemplate())
                || !bean.getTestMethodIndicator().equals(panel.getTestMethodPrefix())
                || bean.isCreateTestIfMissing() != panel.getCreateTestIfMissing()
                || bean.isUnderscoreMode() != panel.getUseUnderscore()
                || bean.isShowFullyQualifiedClassName() != panel.getShowFullyQualifiedClassName()
                || bean.isAutoApplyChangesToTest() != panel.getAutoApplyChangesToTest()
                || bean.isDeletePackageOccurrences() != panel.getDeletePackageOccurrences();
    }

    private boolean equals(List strings1, List strings2) {
        if (strings1.size() != strings2.size()) {
            return false;
        }
        for (int i = 0; i < strings1.size(); i++) {
            if (!strings1.get(i).equals(strings2.get(i))) {
                return false;
            }
        }
        return true;
    }

    public void reset() {
        panel.setCustomMappingStatus(bean.isAllowCustomPackages());
        panel.setCustomPackageMappings(bean.getCustomPackages());
        panel.setTestNameTemplate(bean.getTestNameTemplate());
        panel.setTestMethodPrefix(bean.getTestMethodIndicator());
        panel.setCreateTestIfMissing(bean.isCreateTestIfMissing());
        panel.setUseUnderscore(bean.isUnderscoreMode());
        panel.setShowFullyQualifiedClassName(bean.isShowFullyQualifiedClassName());
        panel.setAutoApplyChangesToTest(bean.isAutoApplyChangesToTest());
        panel.setDeletePackageOccurrences(bean.isDeletePackageOccurrences());
    }

    public ConfigurationBean getState() {
        return bean;
    }

    public void loadState(ConfigurationBean configuration) {
        XmlSerializerUtil.copyBean(configuration, bean);
    }
}

File: trunk/src/scala/org/codehaus/testdox/intellij/config/ConfigurationController.scala
package org.codehaus.testdox.intellij.config

import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.ProjectComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.options.ConfigurationException
import com.intellij.util.xmlb.XmlSerializerUtil
import org.codehaus.testdox.intellij.Icons
import org.codehaus.testdox.intellij.TemplateNameResolver
import org.jetbrains.annotations.NotNull

import javax.swing._

/*
@State(
  name = ConfigurationController.COMPONENT_NAME,
  storages = { @Storage(id = ConfigurationController.COMPONENT_NAME, file = "$PROJECT_FILE$") }
)
*/
class ConfigurationController extends ProjectComponent with Configurable with PersistentStateComponent[Configuration] {

  private var panel: ConfigurationUI = null

  private val configuration = new Configuration()

  setDefaults()

  // BaseComponent ---------------------------------------------------------------------------------------------------

  def initComponent() {}

  def projectOpened() {}

  def projectClosed() {}

  def disposeComponent() {}

  @NotNull
  override val getComponentName = ConfigurationController.COMPONENT_NAME

  private def setDefaults() {
    if (configuration.getCustomPackages == null) {
      configuration.setCustomPackages(List[String]())
    }
    if (configuration.getTestNameTemplate == null) {
      configuration.setTestNameTemplate(TemplateNameResolver.DEFAULT_TEMPLATE)
    }
    if (configuration.testMethodPrefix == null && configuration.testMethodAnnotation == null) {
      configuration.testMethodPrefix = TemplateNameResolver.DEFAULT_PREFIX
    }
    configuration.createTestIfMissing = true
    configuration.autoApplyChangesToTests = true
    configuration.deletePackageOccurrences = true
  }

  // UnnamedConfigurable ---------------------------------------------------------------------------------------------

  @throws(classOf[ConfigurationException])
  def apply() {
    configuration.setCustomPackagesAllowed(panel.customMappingStatus)
    configuration.setCustomPackages(panel.customPackageMappings)
    configuration.setTestNameTemplate(panel.testNameTemplate)
    configuration.createTestIfMissing = panel.createTestIfMissing
    configuration.underscoreMode = panel.useUnderscore
    configuration.setShowFullyQualifiedClassName(panel.showFullyQualifiedClassName)
    configuration.autoApplyChangesToTests = panel.autoApplyChangesToTests
    configuration.deletePackageOccurrences = panel.deletePackageOccurrences

    val prefix = panel.testMethodPrefix
    if (prefix != null && prefix.startsWith("@")) {
      configuration.testMethodPrefix = null
      configuration.testMethodAnnotation = prefix
      configuration.usingAnnotations= true
    } else {
      configuration.testMethodPrefix = prefix
      configuration.testMethodAnnotation = null
      configuration.usingAnnotations = false
    }
  }

  def createComponent(): JComponent = {
    panel = new ConfigurationPanel()
    reset()
    panel.asInstanceOf[ConfigurationPanel]
  }

  def disposeUIResources() { panel = null }

  private[config] def setPanel(panel: ConfigurationUI) { this.panel = panel }

  val getDisplayName = "TestDox"

  val getHelpTopic = ""

  val getIcon = Icons.getIcon(Icons.TESTDOX_ICON)

  def isModified = {
    configuration.getCustomPackagesAllowed != panel.customMappingStatus ||
    configuration.getCustomPackages != panel.customPackageMappings ||
    configuration.getTestNameTemplate != panel.testNameTemplate ||
    configuration.testMethodIndicator != panel.testMethodPrefix ||
    configuration.createTestIfMissing != panel.createTestIfMissing ||
    configuration.underscoreMode != panel.useUnderscore ||
    configuration.getShowFullyQualifiedClassName != panel.showFullyQualifiedClassName ||
    configuration.autoApplyChangesToTests != panel.autoApplyChangesToTests ||
    configuration.deletePackageOccurrences != panel.deletePackageOccurrences
  }

  def reset() {
    panel.customMappingStatus = configuration.getCustomPackagesAllowed
    panel.customPackageMappings = configuration.getCustomPackages
    panel.testNameTemplate = configuration.getTestNameTemplate
    panel.testMethodPrefix = configuration.testMethodIndicator
    panel.createTestIfMissing = configuration.createTestIfMissing
    panel.useUnderscore = configuration.underscoreMode
    panel.showFullyQualifiedClassName = configuration.getShowFullyQualifiedClassName
    panel.autoApplyChangesToTests = configuration.autoApplyChangesToTests
    panel.deletePackageOccurrences = configuration.deletePackageOccurrences
  }

  def getState = configuration

  def loadState(configuration: Configuration) { XmlSerializerUtil.copyBean(configuration, this.configuration) }
}

object ConfigurationController {
  private[config] val COMPONENT_NAME = "TestDoxConfiguration"
}
